### PR TITLE
Blas-1 Copy support (BLAS, CUBLAS)

### PR DESCRIPTION
**Pull request by vqd8a**
Partially address #420. ```Copy``` is needed for Pliris refactoring. This PR is ready for review.

### CODE REVIEW

**Comment by vqd8a** [comment]
Spotcheck on White
```../kokkos-kernels/scripts/test_all_sandia cuda --spot-check --with-cuda-options=enable_lambda --with-tpls=cublas```

```Running on machine: white
Going to test compilers:  cuda/9.2.88 cuda/10.0.130
Testing compiler cuda/9.2.88
Testing compiler cuda/10.0.130
  Starting job cuda-9.2.88-Cuda_OpenMP-release
  PASSED cuda-9.2.88-Cuda_OpenMP-release
  Starting job cuda-10.0.130-Cuda_Serial-release
  PASSED cuda-10.0.130-Cuda_Serial-release
#######################################################
PASSED TESTS
#######################################################
cuda-10.0.130-Cuda_Serial-release build_time=846 run_time=506
cuda-9.2.88-Cuda_OpenMP-release build_time=826 run_time=381
#######################################################
FAILED TESTS
#######################################################
```

```../kokkos-kernels/scripts/test_all_sandia gcc cuda --spot-check --with-tpls=blas```

```Running on machine: white
Going to test compilers:  gcc/6.4.0 gcc/7.2.0 cuda/9.2.88 cuda/10.0.130
Testing compiler gcc/6.4.0
  Starting job gcc-6.4.0-OpenMP_Serial-release
  PASSED gcc-6.4.0-OpenMP_Serial-release
Testing compiler gcc/7.2.0
  Starting job gcc-7.2.0-OpenMP-release
  PASSED gcc-7.2.0-OpenMP-release
  Starting job gcc-7.2.0-Serial-release
  PASSED gcc-7.2.0-Serial-release
  Starting job gcc-7.2.0-OpenMP_Serial-release
  PASSED gcc-7.2.0-OpenMP_Serial-release
Testing compiler cuda/9.2.88
  Starting job cuda-9.2.88-Cuda_OpenMP-release
  PASSED cuda-9.2.88-Cuda_OpenMP-release
Testing compiler cuda/10.0.130
  Starting job cuda-10.0.130-Cuda_Serial-release
  PASSED cuda-10.0.130-Cuda_Serial-release
#######################################################
PASSED TESTS
#######################################################
cuda-10.0.130-Cuda_Serial-release build_time=763 run_time=478
cuda-9.2.88-Cuda_OpenMP-release build_time=830 run_time=381
gcc-6.4.0-OpenMP_Serial-release build_time=420 run_time=370
gcc-7.2.0-OpenMP-release build_time=316 run_time=128
gcc-7.2.0-OpenMP_Serial-release build_time=323 run_time=330
gcc-7.2.0-Serial-release build_time=188 run_time=226
#######################################################
FAILED TESTS
#######################################################
```

```
../kokkos-kernels/scripts/test_all_sandia --spot-check

Running on machine: white
Going to test compilers:  gcc/6.4.0 gcc/7.2.0 ibm/16.1.0 cuda/9.2.88 cuda/10.0.130
Testing compiler gcc/6.4.0
  Starting job gcc-6.4.0-OpenMP_Serial-release
  PASSED gcc-6.4.0-OpenMP_Serial-release
Testing compiler gcc/7.2.0
  Starting job gcc-7.2.0-OpenMP-release
  PASSED gcc-7.2.0-OpenMP-release
  Starting job gcc-7.2.0-Serial-release
  PASSED gcc-7.2.0-Serial-release
  Starting job gcc-7.2.0-OpenMP_Serial-release
  PASSED gcc-7.2.0-OpenMP_Serial-release
Testing compiler ibm/16.1.0
  Starting job ibm-16.1.0-Serial-release
  PASSED ibm-16.1.0-Serial-release  
Testing compiler cuda/9.2.88
  Starting job cuda-9.2.88-Cuda_OpenMP-release
  PASSED cuda-9.2.88-Cuda_OpenMP-release
Testing compiler cuda/10.0.130
  Starting job cuda-10.0.130-Cuda_Serial-release
  PASSED cuda-10.0.130-Cuda_Serial-release
#######################################################
PASSED TESTS
#######################################################
cuda-10.0.130-Cuda_Serial-release build_time=764 run_time=480
cuda-9.2.88-Cuda_OpenMP-release build_time=816 run_time=379
gcc-6.4.0-OpenMP_Serial-release build_time=420 run_time=378
gcc-7.2.0-OpenMP-release build_time=328 run_time=121
gcc-7.2.0-OpenMP_Serial-release build_time=503 run_time=395
gcc-7.2.0-Serial-release build_time=190 run_time=231
ibm-16.1.0-Serial-release build_time=941 run_time=323
#######################################################
FAILED TESTS
#######################################################
```

**Comment by vqd8a** [comment]
Thanks @mhoemmen for your comments.
My main focus in this PR is for TPL BLAS and cuBLAS support for ```copy``` since it is needed in Pliris (dense linear solver). For KokkosKernels implementation, I used the similar one as in ```KokkosBlas::scal``` as a placeholder . I am going to replace it with ```deep_copy``` in the next push. 
However, I would like to keep the BLAS and cuBLAS TPL support, because I noticed that ```deep_copy``` can be slower than BLAS ```copy``` for small vector sizes, and for GPU, it looks like cuBLAS copy is always faster than ```deep_copy```. I attach below some runtime results from my simple tests on White (Pascal60, cuda/10.1.105, openblas/0.3.1/gcc/7.2.0, gcc/7.2.0, OMP_NUM_THREADS=64) with Kokkos::complex<double>: 
[some_runtime_results.txt](https://github.com/kokkos/kokkos-kernels/files/3229326/some_runtime_results.txt)

**Comment by mhoemmen** [comment]
Thanks for the performance results!  I noticed that your test copies from LayoutLeft to LayoutStride.  What is the stride of the LayoutStride View?  Is it stride 1?

@vqd8a wrote:
> However, I would like to keep the BLAS and cuBLAS TPL support, because I noticed that `deep_copy` can be slower than BLAS `copy` for small vector sizes, and for GPU, it looks like cuBLAS copy is always faster than `deep_copy`.

That, to me, says that we need to fix `Kokkos::deep_copy`.  I don't want to stand in your way, but I also don't want to let a performance bug in Kokkos slide past without further inspection. @nmhamster @crtrott

**Comment by kyungjoo-kim** [comment]
I agree with @mhoemmen . The functionality of blas::copy and deep_copy is not different. Having multiple versions of the same functionalities is not a good idea. For the openmp version, the difference is from threading overhead. The blas version seems to be serial implementation. The time increases with the array size and deep copy is more efficient for large array sizes. So, putting this tpl impl with openmp execution space is also problem. In my opinion, application developers (users) should be aware of their problem granularity and if the problem is smaller than a certain threshold, they should use serial space instead of openmp space. For the cuda case, it is a more or less tuning issue in md range policy (matrix case) that should be addressed in kokkos.

**Comment by vqd8a** [comment]
@mhoemmen I just randomly used stride =10 in my test.

In my Pliris code, using ```deep_copy``` is even worse. The total performance when using BLAS ```copy``` is 6.6x faster than when using ```deep_copy```, although this number is from a small problem size. I will test with larger problem sizes too.   

I agree that we should inspect why Kokkos::deep_copy performs that way in the Kokkos side.

I also think it is not bad to allow users to choose to use ```deep_copy```, BLAS ```copy``` or cuBLAS ```copy```.

**Comment by mhoemmen** [comment]
> I also think it is not bad to allow users to choose to use `deep_copy`, BLAS `copy` or cuBLAS `copy`.

I disagree.  If `Kokkos::deep_copy` is slower than either of these for a very common case, then we need to fix `Kokkos::deep_copy`.  We don't need `N >> 1` ways of doing the same thing, some of which are mysteriously slower than others.

**Comment by vqd8a** [comment]
@mhoemmen I somehow agree with you. But I am not quite clear why we have KokkosKernels implementation for BLAS functions (dot, abs, scal, gemv, gemm,...) and also have BLAS, cuBLAS TPL support for these.

**Comment by mhoemmen** [comment]
@vqd8a wrote:
> But I am not quite clear why we have KokkosKernels implementation for BLAS functions (dot, abs, scal, gemv, gemm,...) and also have BLAS, cuBLAS TPL support for these.

1. There's a difference between implementation and interface.  If the fastest way to implement `Kokkos::deep_copy` for certain types is to call cuBLAS, then we should do that.  However, I don't think we should make multiple interfaces available to users.

2. In the non-CUDA case, the BLAS' copy is faster than `Kokkos::deep_copy` for short vectors because the BLAS doesn't launch an OpenMP parallel region.  This just means that `Kokkos::deep_copy` also should not launch an OpenMP parallel region for short vectors.

File: scripts/test_all_sandia [hunk 1]
```diff
@@ -296,7 +296,7 @@ elif [ "$MACHINE" = "white" ]; then
   module load netlib/3.8.0/gcc
   export BLAS_LIBRARIES="${BLAS_ROOT}/lib/libblas.a"
   # blas is compiled by gcc so use single underscore and power8 use return complex
-  KOKKOS_OPTIONS="eti-only,blas-mangle_,blas-return-complex"
+  KOKKOS_OPTIONS="--with-options=eti-only,blas-mangle_,blas-return-complex"
```
[the remark this answers is not in the document]
Comment (crtrott) [reply scripts/test_all_sandia]: I am not sure. I think it is not on.

[the remark this answers is not in the document]
Comment (crtrott) [reply scripts/test_all_sandia]: But we need to test that reasonably regularly. That said this change doesn't seem to make sense. If it doesn't add the "--with-options" correctly we need to fix that in the script.

**Pull request closed without merging**